Invalid API endpoint exception improvement - MAILPOET-6022

triple0t opened the pull request:
## Description

Invalid API endpoint exception improvement

## Code review notes

_N/A_

## QA notes

Please check Jan's [note](https://github.com/mailpoet/mailpoet/pull/5592#issuecomment-2133383764) below
You can rename or delete one of the [JSON API endpoints](https://github.com/mailpoet/mailpoet/tree/605664228ee1bd337ac5024eec4526778a5ed7fa/mailpoet/lib/API/JSON/v1)

## Linked PRs

_N/A_

## Linked tickets

[MAILPOET-6022](https://mailpoet.atlassian.net/browse/MAILPOET-6022)

## After-merge notes

_N/A_

## Tasks

- [ ] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [ ] I added sufficient test coverage
- [ ] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-6022]: https://mailpoet.atlassian.net/browse/MAILPOET-6022?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

triple0t committed:
Invalid API endpoint exception improvement

MAILPOET-6022

triple0t committed:
Fix failing test

MAILPOET-6022

lysyjan reviewed (commented):
@triple0t Thank you for this change.
I tried to replicate this behavior briefly, but the message was dismissed after quite a short time, so the link was useless for me because I could not click or copy the link. Do you think we could find a way to make the link clickable? I am afraid that in this state, the change will not have the expected impact. What do you think about it?

Also, adding replication steps for QA would be great.

triple0t commented:
Thanks for the review, @lysyjan 

> I tried to replicate this behavior briefly, but the message was dismissed after quite a short time, so the link was useless for me because I could not click or copy the link.

Hmm, I'm not returning the message as an HTML because not all places on the Frontend accept such error messages.
There are probably a few I could find and update, but there are way more I'll miss (here is an [example](https://github.com/search?q=repo%3Amailpoet%2Fmailpoet%20.fail((response)%20%3D%3E&type=code)).
I decided on this approach to be backward compatible. 

We definitely don't want the message to be easily dismissed though.
Can you send me the link to the component you observed this on?

> Also, adding replication steps for QA would be great.

Thanks for the suggestion. 
I could replicate this error by intentionally throwing an error in the code.
Do you have an easy way to allow QA to test this?

Otherwise, I was planning on creating a .zip file with the modified code, which shows/throws the exception.

lysyjan commented:
> Hmm, I'm not returning the message as an HTML because not all places on the Frontend accept such error messages.
There are probably a few I could find and update, but there are way more I'll miss (here is an [example](https://github.com/search?q=repo%3Amailpoet%2Fmailpoet%20.fail((response)%20%3D%3E&type=code)).
I decided on this approach to be backward compatible.

I'm not sure about this approach because it is not user-friendly. Copying the URL from the error message offers more space for a user's mistake. What do you think @pavel-mailpoet about it? I'm asking you because you created this ticket.

> We definitely don't want the message to be easily dismissed though.
Can you send me the link to the component you observed this on?


@triple0t I tested this behavior by renaming one of the JSON API endpoints, for example, from `Newsletters.php` to `Newsletter.php`, and then I visited the newsletters listing page. So, I don't know which component it was.

pavel-mailpoet commented:
> I'm not sure about this approach because it is not user-friendly. Copying the URL from the error message offers more space for a user's mistake. What do you think @pavel-mailpoet about it? I'm asking you because you created this ticket.

If it is easy, we should definitely go for the user-friendly approach. That is what we are trying. Help users so that they don't need to contact us.

triple0t committed:
Add html to the error message

MAILPOET-6022

triple0t committed:
Use MailPoet API Error notice when possible.

This will ensure the error message is well displayed including any additional html content

MAILPOET-6022

triple0t commented:
I've updated the PR. Kindly help take a look @lysyjan

lysyjan commented:
@triple0t When I tried to test the new error message, I saw a non-clickable link:
<img width="1544" alt="Screenshot 2024-05-29 at 11 27 50" src="https://github.com/mailpoet/mailpoet/assets/13644846/d6f0d9a2-ef4f-4bd7-b023-d2c2cbf11119">

Did I do something wrong? Also, the message was automatically dismissed after several seconds. Of course, I compiled the latest JS from your branch.

triple0t committed:
Add support for MailPoet API Error notice in Frontend JS files.

This will ensure the error message is well displayed including any additional html content

MAILPOET-6022

triple0t committed:
Fix lint error

MAILPOET-6022

triple0t commented:
Updated @lysyjan 

Thanks for the feedback

lysyjan commented on mailpoet/assets/js/src/context/use-notices.jsx:
```diff
@@ -75,13 +75,15 @@ export const useNotices = () => {
                       '<a> https://kb.mailpoet.com/article/258-re-installing-updating-the-plugin-via-ftp </a>',
                     ),
                     {
-                      // eslint-disable-next-line jsx-a11y/anchor-has-content, jsx-a11y/control-has-associated-label
```
Just a note that this comment didn't work because it should have been at the line below.

triple0t replied:
This was fixed here https://github.com/mailpoet/mailpoet/pull/5592/commits/4a39e3b5ce1f27a931a31659367ed13e38f31069

lysyjan commented on mailpoet/assets/js/src/context/use-notices.jsx:
```diff
@@ -46,12 +48,69 @@ export const useNotices = () => {
+                        'The plugin has encountered an unexpected error. Please reload the page. If that does not help, re-install the MailPoet Plugin. See: %s for more information',
+                        'mailpoet',
+                      ),
+                      '<a> https://kb.mailpoet.com/article/258-re-installing-updating-the-plugin-via-ftp </a>',
```
- The text in the link element should be different from the ticket specification. Or is there a reason why you used this approach?
- Also, we usually use the <link> mark here, which can be a part of the translation string, and translators already know it. Maybe we should stay with this approach.
- Is there some reason for the redundant spaces around the text inside the tag?

triple0t replied:
> The text in the link element should be different from the ticket specification. Or is there a reason why you used this approach?

Uh? Why should the text in the link element be different?

> Also, we usually use the mark here, which can be a part of the translation string, and translators already know it. Maybe we should stay with this approach.

I'm lost. Are you suggesting we use <mark> instead of <a> ?  We need the <a> for the interpolation.

> Is there some reason for the redundant spaces around the text inside the tag?

Not really. It can/will be removed.

lysyjan replied:
> Uh? Why should the text in the link element be different?
> I'm lost. Are you suggesting we use instead of ? We need the for the interpolation.

@triple0t The ticket specification contains the following sentence:
`The plugin has encountered an unexpected error. Please reload the page. If that doesn't help [link]re-install the MailPoet Plugin.[/link]`,
but you used: `The plugin has encountered an unexpected error. Please reload the page. If that does not help, re-install the MailPoet Plugin. See: %s for more information`.
Displaying the target URL can make the text less readable. I don't think we use this approach anywhere.
Also, when we use string replacing via `sprintf` with the `%s` specifier, we should add a comment for translators. [Here](https://github.com/mailpoet/mailpoet/blob/c143da4ccfe266beb6091ce0ede742bad2ff639c/mailpoet/assets/js/src/automation/listing/components/menu/delete.tsx#L47-L51) is an example.

Nevertheless, If I add this message, I would try to follow the ticket specification exactly, and it could look like this:
```
createInterpolateElement(
  __(
    'The plugin has encountered an unexpected error. Please reload the page. If that does not help, <link>re-install the MailPoet Plugin.</link>,
    'mailpoet',
  ),
  {
    link: (
    <a
      href="https://kb.mailpoet.com/article/258-re-installing-updating-the-plugin-via-ftp"
      target="_blank"
      rel="noopener noreferrer"
    >
      &nbsp;
    </a>
    ),
  },
)
```

Maybe we can display the whole URL in the text, but in my opinion, it should be at the end of the sentence and not in the middle, and we would need to add a comment for translators.

triple0t replied:
Noted. 
Fixed here https://github.com/mailpoet/mailpoet/pull/5592/commits/656a66299bd1b2c85449b60b394983bdfab88455
Thanks

lysyjan reviewed (commented):
@triple0t Thanks for the update.

Because this is very complex issue that we are trying to solve. I was testing it again and maybe more than is from a developer expected, but I found some incosistencies for some pages. For example, the Automations page has a different behavior:
<img width="1470" alt="Screenshot 2024-05-30 at 13 22 25" src="https://github.com/mailpoet/mailpoet/assets/13644846/49bb13e6-d6de-4f57-811e-5440d2642d80">

Error message on the Segments page is not rendered properly:
<img width="1300" alt="Screenshot 2024-05-30 at 13 23 46" src="https://github.com/mailpoet/mailpoet/assets/13644846/40692c63-ea48-44c0-80e1-1e1010e554ed">

The Subscribers and Lists pages end with an infinite loop of error messages, but here I'm not sure if we should create a new ticket for it or address it here.
<img width="1290" alt="Screenshot 2024-05-30 at 13 25 46" src="https://github.com/mailpoet/mailpoet/assets/13644846/92dfd3b7-f8ee-4c3b-bb1b-fad9b70bd317">

triple0t committed:
Use MailPoet error notices in Dynamic Segments

MAILPOET-6022

triple0t committed:
Fix infinite error messages notice.

This is because we call `getItems()` in a few places already and this `notices.apiError` causes a render which causes `getItems()` to re-render, and so on.

MAILPOET-6022

triple0t commented:
Thanks for the in-depth review, @lysyjan 

I have addressed the issues you noted, except for the automation one.

Automation uses WP JSON directly and not MP REST endpoint. Configuring that to work with the use case in this PR will require some effort and a few refactors. I don't want to do something out of scope for this ticket, so I'll leave it as is.

lysyjan commented:
@triple0t Thanks for fixing the issues. I think we can create a ticket for better handling errors on the Automation page.

Please, take a look at the last comment related to the message format.

triple0t committed:
Update the notice message to fix consistency

MAILPOET-6022

Aschepikov reviewed (approved):
GJ!
